app/models/classification.rb

```diff
@@ -10,13 +10,14 @@ class Classification < ApplicationRecord
   before_destroy :delete_tags_and_entries
 
   validates :description, :presence => true, :length => {:maximum => 255}
-  validates :description, :uniqueness => {:scope => [:parent_id]}, :if => proc { |c|
-    cond = ["parent_id = ? AND description = ?", c.parent_id, c.description]
-    unless c.new_record?
-      cond.first << " AND id != ?"
-      cond << c.id
-    end
-    c.class.in_region(region_id).exists?(cond)
+  validates :description, :uniqueness => { :scope => %i(parent_id tag_id) }, :if => proc { |c|
```

Comment:
@AlexanderZagaynov can you describe why we need this? Amazon works without these changes, right?

Reply:
Description uniqueness previously was validated within `parent_id` + `in_region` scope. `If` condition was probably created because of bizarre `in_region` scope, which differentiated through `id`s range, not through some designated field.
My change just adds `tag_id` to uniqueness scope, and introduces slight `if` proc code refactoring using standard AR methods.

Reply:
Not to defend the existing validation (:wink: I've stared at this for 2 years and I'm still not sure how to read it! :worried:), but please explain the change.
Wouldn't scoping on `tag_id` too allow as many Classifications with same description (but different Tag) as you want, effectively losing uniqueness?

Reply:
Hi @cben to answer your question about uniqueness scope we need to find out first why do we need that uniqueness in the whole. I can't say that. But I can say that since `tag` should be unique for each provider, and `classification` should refer to the tag - it's impossible to have mapping for both `Amazon` and `Azure` to the same classification name without this change.

Reply:
Aha. Not sure I'm understand right.  You mean we might want to simultaneously map say:
`foo`, AzureVm -> "Foo" classification, `/managed/azure:vm:foo`
`foo`, AmazonImage -> "Foo" classification, `/managed/amazon:template:foo` (I'm probably off in exact type & resulting tag name)
?

So indeed that would allow user to see in UI:
`foo=bar` azure vm tag mapped to a "Foo : bar" MIQ tag,
`foo=baz` amazon image tag mapped to a "Foo : baz" MIQ tag.

However, that's a very misleading situation.  It'd *look* like tags in same category, but actually there'd be 2 distinct categories with same name!  Then when you'd want say a policy or report by tag in Foo category, you'd see 2 Foo categories, looking exactly the same, and would have to choose one...

----

If you really want to achieve simultaneous several types -> one category mappings, we should modify UI to not include mapping type in tag name, and to allow several ContainerLabelTagMapping to point to same category tag_id.  (Mapper has no problem with that.)
The problem doing that in UI is "garbage collection" of categories — currently I delete category with the mapping, would have to do a check to see if it's still used...
BTW any change in how we construct categories probably needs data migration.

But I want to question do we really need this?
If one wants 2 things mapped to one category, just do an "\<All>" mapping.

Reply:
do we need this for this PR? I don't think so, right? Lets separate this change to another PR and bring @Fryguy and others, since this has broader impact than just tag mapping (I don't know what the impact ca be)

Reply:
@Ladas are you asking me? If so, my answer is: yes, definitely.

Reply:
@AlexanderZagaynov yeah. Yes for we can separate it to another PR? :-)

Reply:
No, we can't. It will not work without this change.

Reply:
Can you please explain full use case that doesn't work with current validation? (what mappings and what classifications are involved?)

Reply:
Ok, probably my https://github.com/ManageIQ/manageiq/pull/17212/files#r181664339 wasn't clear enough. I did draw the picture to illustrate what's happeing inside.
Without that change we will not be able to add two classifications with the same description for a different providers, and all the labels will stick to the single mapping entry only, even from the different providers.
![labeltagmapping](https://user-images.githubusercontent.com/7778414/38812401-7b3adf1c-418c-11e8-8899-4d01974a1325.png)

Reply:
@cben I guess you gave a weighty arguments here, so I'm going to roll that validation scope change back 
(cc @Ladas)